pyfr/integrators/dual/pseudo/multip.py

```diff
@@ -0,0 +1,292 @@
+# -*- coding: utf-8 -*-
+
+from collections import defaultdict
+import itertools as it
+import re
+
+from pyfr.inifile import Inifile
+from pyfr.integrators.dual.pseudo.base import BaseDualPseudoIntegrator
+from pyfr.integrators.dual.pseudo.pseudocontrollers import BaseDualPseudoController
+from pyfr.integrators.dual.pseudo.pseudosteppers import BaseDualPseudoStepper
+from pyfr.util import memoize, proxylist, subclass_where
+
+
+class DualMultiPIntegrator(BaseDualPseudoIntegrator):
+    def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg,
+                 tcoeffs):
+        self.backend = backend
+
+        sect = 'solver-dual-time-integrator-multip'
+
+        # Get the solver order and set the initial multigrid level
+        self._order = self.level = cfg.getint('solver', 'order')
+
+        # Get the multigrid cycle
+        self.cycle, self.csteps = zip(*cfg.getliteral(sect, 'cycle'))
+        self.levels = sorted(set(self.cycle), reverse=True)
+
+        if max(self.cycle) > self._order:
+            raise ValueError('The multigrid level orders cannot exceed '
+                             'the solution order')
+
+        if any(abs(i - j) > 1 for i, j in zip(self.cycle, self.cycle[1:])):
+            raise ValueError('The orders of consecutive multigrid levels can '
+                             'only change by one')
+
+        if self.cycle[0] != self._order or self.cycle[-1] != self._order:
+            raise ValueError('The multigrid cycle needs to start end with the '
+                             'highest (solution) order ')
+
+        mgsect = 'solver-dual-time-integrator-multip'
+        sect = 'solver-time-integrator'
+
+        # Multigrid pseudo-time steps
+        dtau = cfg.getfloat(sect, 'pseudo-dt')
+        dtauf = cfg.getfloat(mgsect, 'pseudo-dt-fact', 1.0)
+        self.dtaus = {l: dtau*dtauf**(self._order - l) for l in self.levels}
```

Comment:
I don't think we need this.  We can simply generate the dtau we need in the main for loop.